lib/timers.js

```diff
@@ -440,13 +452,20 @@ function rearm(timer) {
 
 
 const clearTimeout = exports.clearTimeout = function(timer) {
-  if (timer && (timer[kOnTimeout] || timer._onTimeout)) {
+  if (!timer) return;
+
+  if (timer[kOnTimeout] || timer._onTimeout) {
+    if (typeof timer._onTimeout !== 'function') {
+      promiseResolve(timer._onTimeout, undefined);
+    }
     timer[kOnTimeout] = timer._onTimeout = null;
     if (timer instanceof Timeout) {
       timer.close(); // for after === 0
     } else {
       unenroll(timer);
     }
+  } else if (isPromise(timer)) {
```

Comment:
I have to agree with @vkurchatkin on this. Until there is a standard way of canceling promises, we likely shouldn't do this.